### DISCUSSION TITLE
Equivalent of "maximum" quality export in Photos

### DISCUSSION THREAD

**Discussion by kitschpatrol** [RhetTbull/osxphotos#1522] [General]
It seems like Photos.app implements some additional plumbing to re-render the stack of image adjustments at a higher quality when TIFF, PNG, or "Maximum" quality JPEG is requested via the export command GUI:

![image](https://github.com/RhetTbull/osxphotos/assets/194164/06893ec9-932b-4a8e-8baf-7b90ddde4ebe)

However, as far as I can find, all the available programmatic export techniques seem to top out at the equivalent of the "High" quality Photos.app JPEG export, sometimes resulting in non-trivial compression artifacts in the resulting images.

Seems like there's no way to specify export format or quality via AppleScript, and I tried writing a [quick Swift app](https://gist.github.com/kitschpatrol/4754b5cd67ba25dc41d7370855f9f193) using PhotoKit directly to poke around and see if there were any options to request additional quality. I found none.

Interestingly PowerPhotos exposes additional export options (TIFF, "Maximum" quality JPEGs, etc.), but these don't seem to yield the same output as the equivalent options Photos.app. For example, a TIFF export comes with JPEG artifacts reminiscent of the "High" quality export from Photos.app.

I tested across a couple different export techniques for an image edited in Photos.app which was originally a 6MB jpeg out of the camera. The results are summarized below, with many landing at 497,748 bytes, suggesting the export is a straight copy from the edited JPEG stored in the Photos library filesystem:

| Application                                  | Quality specified   | Size in Bytes | Artifacts |
| -------------------------------------------- | -------------------| -------------:| ---------------------- |
| osxphotos                                    | N/A                 | **497,748**       | Yes                    |
| osxphotos --convert-to-jpeg                  | 1.0 (default)       | **497,748**       | Yes                    |
| osxphotos --use-photos-export                | N/A                 | 751,267       | Yes                    |
| osxphotos --use-photos-export --use-photokit | N/A                 | **497,748**       | Yes                    |
| Photos.app export GUI                            | JPEG High           | 468,230       | Yes                    |
| Photos.app export GUI                            | JPEG Maximum        | 5,017,605     | No                     |
| Photos.app export GUI                                   | TIFF                | 19,603,750    | No                     |
| Photos.app library filesystem                    | N/A                | **497,748**       | Yes                    |
| PowerPhotos export GUI                                  | JPEG High        | 508,669     | Yes                    |
| PowerPhotos export GUI                                  | JPEG Maximum        | 1,415,768     | Yes                    |
| PowerPhotos export GUI                                  | TIFF                | 19,601,382    | Yes (!?)               |
| Retrobatch "Photos Export"                                  | N/A | **497,748**       | Yes                     |
| [DIY Swift PhotoKit application](https://gist.github.com/kitschpatrol/4754b5cd67ba25dc41d7370855f9f193)               | highQualityFormat   | **497,748**       | Yes                    |
| Automator Export Media Items               | N/A   | 751,267       | Yes                    |
| Photomator export GUI               | JPEG 100%   | 3,079,726       | Yes                    |
| Photomator export GUI               | TIFF   | 3,862,480       | Yes                    |
| [photohack](https://github.com/neilpa/photohack)               | highQualityFormat   | 3,066,131       | Yes                    |

I realize the relationship between file size and image quality is not guaranteed, but on inspection there are subtle but perceptible differences between the ~500K and ~5MB JPEG exports.

This is in no way the fault of osxphotos, and of course exporting the original image will preserve full quality regardless — but I'm curious if anyone else has noticed this discrepancy between "High" and "Maximum" quality exports of edited images, and if there's any sort of work around or option that can automate exports of a quality equivalent to that available in the (tragically cumbersome) Photos.app GUI.

**Comment by RhetTbull** [comment]
Interesting analysis -- thanks for sharing. Currently osxphotos exports only the cached edited image stored in the library (unless the image is missing and you use `--download-missing` or as you tried, you use `--use-photos-export`.  I've never worried too much about this because I always export the original image too and I could recreate the edit if needed. 

It would definitely be useful to have a way to export the "Maximum" quality image as you noted. I have some code where I've been playing with the private Photos API (which I hope to eventually turn into a complete rewrite of osxphotos). I'll take a look and see if there are any clues to doing this.

**Reply by RhetTbull** [reply]
Here's some data from my testing. For me, `osxphotos export --use-photos-export` exported the "maximum size JPEG" but labeled it as a HEIC so that may be a bug in OSXPhotos (or it could be a bug in the AppleScript interface for Photos...will need to investigate).

I don't know what you mean by artifact so didn't look for those. The test image was an original HEIC and the exported images were also HEIC unless otherwise specified below:

```
Photos
unmodified original: 2,880,850
edited, JPEG, high: 3,800,977
edited, JPEG, maximum: 13,139,449
edited on filesystem, HEIC: 3,185,582

Photoscript:
edited (jpeg): 13,139,449

OSXPhotos:
unmodified original: 2,880,850
edited (heic): 3,185,582

OSXPhotos --convert-to-jpeg, 1.0 quality:
unmodified original: 9,903,013
edited (jpeg): 10,566,039

OSXPhotos --use-photos-export
unmodified original: 2,880,850
edited (OSXPhotos exported with .heic but it's actually a JPEG): 13,139,449

OSXPhotos --use-photos-export --use-photokit
unmodified original: 2,880,850
edited: 3,185,582

PhotoKit:
PHImageRequestOptionsVersionOriginal: 2,880,850
PHImageRequestOptionsVersionCurrent: 3,185,582
PHImageRequestOptionsVersionUnadjusted: 2,880,850
```

**Reply by RhetTbull** [reply]
For the `--use-photos-export` option, OSXPhotos uses my [PhotoScript](https://github.com/RhetTbull/PhotoScript) package which uses AppleScript to do the export. At least in my tests, this appears to export the maximum size photo. Save the following to `export_selected.py` and run it via `osxphotos run export_selected.py  /path/to/export/`.  What do you get?

```python
"""Export selected photos from Photos library to a folder"""

import click
import photoscript

from osxphotos.cli.click_rich_echo import rich_click_echo as echo


@click.command()
@click.argument("DESTINATION", type=click.Path(exists=True, file_okay=False))
def export_selected(destination):
    """Export selected photos from Photos library to a folder"""
    photos = photoscript.PhotosLibrary().selection
    if not photos:
        echo("No photos selected")
        return
    for photo in photos:
        if original := photo.export(destination, original=True):
            echo(
                f"Exported original version of {photo.filename} ({photo.uuid}) to {original[0]}"
            )
        else:
            echo(f"Failed to export original version of {photo.filename} ({photo.uuid})")

        if current := photo.export(destination, original=False):
            echo(
                f"Exported current version of {photo.filename} ({photo.uuid}) to {current[0]}"
            )
        else:
            echo(f"Failed to export current version of {photo.filename} ({photo.uuid})")


if __name__ == "__main__":
    export_selected()

```

**Reply by kitschpatrol** [reply]
Thanks so much for digging into this.

By "artifacts" I just mean signs of JPEG compression / quantization, e.g. banding in color gradients and that sort of thing which (at least in my tests) are more apparent in the `osxphotos` export than in a direct "maximum quality" export from the Photos GUI. I'll start running SSIM on the output instead to get something a bit more quantitative.

Using your provided `export_selected.py` script, I'm seeing the same behavior as the `--use-photos-export` example from my original post (as expected since I understand `--use-photos-export` calls the photoscript-powered export path.)

The original file exports identically whether via GUI or `export_selected.py`, but the edited / current image is still differing from a "maximum" quality export from the Photos GUI:

The SSIM / DSSIM / PSNR measurements are vs. the "maximum" quality GUI export.

| Export strategy                   | File Size | DSSIM      | PSNR      | SSIM     |
| ------------------------- | --------- | ---------- | --------- | -------- |
| Photos.app GUI "Maximum" JPEG | 5,017,605 | 0          | 0         | 1        |
| `export_selected.py` current version        | 751,267   | 0.00095077 | 33.461056 | 0.908169 |

And that's really interesting about HEIC vs. JPEG edits, in my test case the original image is a JPEG, so maybe that's related. Or maybe there's some difference in the Photos.app GUI state between our machines that's affecting the AppleScript-driven export?

**Reply by RhetTbull** [reply]
I just tried with a non-HEIC image. The PhotoScript script exports the same maximum quality JPEG (7.7MB) as the GUI does but osxphotos exports a 1.9MB version (the cached version). I'll dig around to see if there's some setting in Photos that may be driving this. I thought it was maybe "last used setting" but I confirmed the script exports the maximum size even if last used setting in the GUI was "low".  I'll keep investigating.  I'm on Ventura 13.5.1 so perhaps its an OS-specific setting.

**Reply by RhetTbull** [reply]
Update: some progress using PhotoKit's `requestimageforasset:targetSize:contentMode:options:resultHandler:` with `PHImageManagerMaximumSize` and `PHImageRequestOptionsResizeModeExact`. The resulting JPEG is 5.9MB vs 1.9MB cached version but still not the full 7.7MB "maximum quality".  Will keep tinkering.

**Reply by RhetTbull** [reply]
> The SSIM / DSSIM / PSNR measurements are vs. the "maximum" quality GUI export.

What tool do you use to measure this?

**Reply by kitschpatrol** [reply]
Thanks, those PhotoKit options sound really promising.

I can confirm that I also haven't found any Photos.app GUI settings that seem to change the behavior of these other tools.

Version-wise I'm on Sonoma 14.4.1 and Photos.app 9.0 (646.2.101).

For the similarity measurements, there are multiple implementations of each algorithm out there, in my tests I'm using:

- https://kornel.ski/dssim (`dssim <image1> <image2>`)
- FFmpeg's PSNR implementation (`ffmpeg -i <image1> -i <image2> -filter_complex psnr -f null -`)
- FFmpeg's SSIM implementation (`ffmpeg -i <image1> -i <image2> -filter_complex ssim -f null -`)

**Reply by RhetTbull** [reply]
Correction: it was actually [requestimageforasset:targetSize:contentMode:options:resultHandler:](https://developer.apple.com/documentation/photokit/phimagemanager/1616964-requestimageforasset) that I was able to use to export a larger version with those options.

**Reply by RhetTbull** [reply]
I've not yet been able to figure out how to export the "maximum quality" version but a downside to [requestimageforasset:targetSize:contentMode:options:resultHandler:](https://developer.apple.com/documentation/photokit/phimagemanager/1616964-requestimageforasset) is that doing it this way strips the metadata. There may be a way to add it back in but haven't had time to play with this.

**Reply by kitschpatrol** [reply]
Interesting, yes I tried out [requestimageforasset](https://developer.apple.com/documentation/photokit/phimagemanager/1616964-requestimageforasset) and am seeing the same results you are reporting, with a slightly larger file and stripped metadata. (I think the metadata loss is expected since it's not present in the NSImage representation, but my Cocoa is a bit rusty.)

One other variable that's come to my attention is the color profile... it seems like any color profile change that triggers a reprocessing of the pixel data (e.g. a conversion instead of a simple profile assignment) will have a dramatic effect on the byte size of a "maximum" quality JPEG. ~I think this might have been affecting my original test report from a lingering setting in the Photos.app GUI, and is responsible for the difference between the 751k and the 5MB exports in my original table — so, very sorry for the confusion there — but this explains why you were getting equivalent file sizes from `export_selected.py` when I was not.~ (Update: Can't reproduce this significant of a variation.)

Also, rethinking my testing protocol, since a lossless TIFF export is available in the Photos.app GUI, and since [requestimageforasset](https://developer.apple.com/documentation/photokit/phimagemanager/1616964-requestimageforasset) is returning an NSImage, I'm saving that down to a TIFF as well to eliminate a potential JPEG compression pass. But still not getting equivalent output between the [requestimageforasset](https://developer.apple.com/documentation/photokit/phimagemanager/1616964-requestimageforasset) and Photos.gui export versions. But the image returned from [requestimageforasset](https://developer.apple.com/documentation/photokit/phimagemanager/1616964-requestimageforasset) also _doesn't_ seem like an exact match for the cached edited image in the Photos library, suggesting that some processing or resampling of the edit stack is taking place, but I'm not 100% sure of this yet. Will keep trying permutations.

**Reply by kitschpatrol** [reply]
Digging into Photos.app a bit with Instruments, it seems to be calling a few methods of interest during the export process, things like:

`exportimageToURL:composition:options:completion:` on `PICompositionExporter` from the private PhotoImaging.framework. This seems to be associated with calls to Metal-backed rendering methods. Also seeing calls to [`requestContentEditingInput(with:completionHandler:)`](https://developer.apple.com/documentation/photokit/phasset/1624061-requestcontenteditinginput) on `PHAsset`.

I'm not seeing these same methods called when using the public PhotoKit image request methods. This vaguely supports the hypothesis that Photos.app is re-rendering the adjustment stack on export via the GUI in a way that doesn't quite match what we're getting from PhotoKit.

I tried playing around with `requestContentEditingInput` to see if I could use this to trigger a re-render, but no luck so far, saving an image directly from the `fullSizeImageURL` value provided by the editing request seems to match the output from `requestImage(for:targetSize:contentMode:options:resultHandler:)`.

**Reply by RhetTbull** [reply]
Great detective work! I'm playing around with `exportimageToURL:composition:options:completion:` to see if I can reverse engineer the call. I'm able to load the class and call the method but not successfully yet.  I'll play around with Instruments when I get a chance to see if I can determine the full sequence of methods that needs to be called.

**Reply by kitschpatrol** [reply]
Thanks for the update! `exportimageToURL:composition:options:completion:` definitely seems promising, but the `composition:` argument looks tricky. Interested to hear what you figure out.

I'm pretty deep down the rabbit hole at this point 😅, so to claw back some sanity I put together a script to automate the comparison of different export strategies.

Here are some tables output by the script:

---

Image ID: 0CC54CB1-EE02-4230-80B1-D8CB8CF229CE

| Export Method                                | Format | Profile    |   Size | EXIF Tags |  SSIM |
| :------------------------------------------- | :----- | :--------- | -----: | --------: | ----: |
| File system copy (Original)                  | CR2    | Display P3 | 26136k |       332 |    NA |
| File system copy                             | JPEG   | sRGB       |  2939k |        97 | 0.997 |
| `osxphotos export`                           | JPEG   | sRGB       |  2939k |        97 | 0.997 |
| `osxphotos export --photo-kit`               | JPEG   | sRGB       |  2939k |        97 | 0.997 |
| `osxphotos export --use-photos-export`       | JPEG   | sRGB       |  3383k |        99 | 0.993 |
| Photos GUI Export JPEG High                  | JPEG   | sRGB       |  2816k |        99 | 0.991 |
| Photos GUI Export JPEG Maximum                   | JPEG   | sRGB       |  7355k |        99 | 0.998 |
| Photos GUI Export PNG                        | PNG    | sRGB       | 15838k |        98 | 1.000 |
| PhotoKit `requestImage...`                   | PNG    | sRGB       | 14832k |        28 | 0.997 |
| PhotoKit `requestImageDataAndOrientation...` | PNG    | sRGB       | 14870k |        28 | 0.997 |

---

Image ID: 77758382-025A-446E-91C6-88A0BCAFDA91

| Export Method                                | Format | Profile    |  Size | EXIF Tags |  SSIM |
| :------------------------------------------- | :----- | :--------- | ----: | --------: | ----: |
| File system copy (Original)                  | JPEG   | sRGB       | 5960k |       254 |    NA |
| File system copy                             | JPEG   | Display P3 |  494k |       122 | 0.877 |
| `osxphotos export`                           | JPEG   | Display P3 |  494k |       122 | 0.877 |
| `osxphotos export --photo-kit`               | JPEG   | Display P3 |  494k |       122 | 0.877 |
| `osxphotos export --use-photos-export`       | JPEG   | Sharing    |  746k |       128 | 0.904 |
| Photos GUI Export JPEG High                  | JPEG   | sRGB       |  465k |       105 | 0.879 |
| Photos GUI Export JPEG Maximum                   | JPEG   | sRGB       | 4987k |       105 | 0.992 |
| Photos GUI Export PNG                        | PNG    | sRGB       | 5958k |       101 | 1.000 |
| PhotoKit `requestImage...`                   | PNG    | Display P3 | 3584k |        53 | 0.747 |
| PhotoKit `requestImageDataAndOrientation...` | PNG    | Display P3 | 3437k |        59 | 0.877 |

---

Image ID: B60D9534-6ADE-4BD9-83EC-A95C8DEBFB8F

| Export Method                                | Format | Profile    |   Size | EXIF Tags |  SSIM |
| :------------------------------------------- | :----- | :--------- | -----: | --------: | ----: |
| File system copy (Original)                  | JPEG   | sRGB       | 11171k |        97 |    NA |
| File system copy                             | JPEG   | Display P3 | 10950k |       120 | 0.989 |
| `osxphotos export`                           | JPEG   | Display P3 | 10950k |       120 | 0.989 |
| `osxphotos export --photo-kit`               | JPEG   | Display P3 | 10950k |       120 | 0.989 |
| `osxphotos export --use-photos-export`       | JPEG   | sRGB       | 10705k |       111 | 0.985 |
| Photos GUI Export JPEG High                  | JPEG   | sRGB       | 10701k |       111 | 0.986 |
| Photos GUI Export JPEG Maximum                   | JPEG   | sRGB       | 33624k |       111 | 0.996 |
| Photos GUI Export PNG                        | PNG    | sRGB       | 66691k |       111 | 1.000 |
| PhotoKit `requestImage...`                   | PNG    | Display P3 | 59068k |        53 | 0.991 |
| PhotoKit `requestImageDataAndOrientation...` | PNG    | Display P3 | 60700k |        53 | 0.989 |

---

These results mostly show what we've seen before, with "fatter" files coming out of the GUI export, and multiple other export paths simply copying the cached edit file straight from the `.photoslibrary` folder . Some of the color profile changes vs. the original image are unexpected, I'm trying to figure out the logic there. (Maybe editing in Photos converts file to Display P3 sometimes.) The SSIM values are a bit arbitary since I (somewhat subjectively) set the Photos GUI PNG as the benchmark, but they're still useful for quantifying variation across the export strategies.

In the course of automating this, I wrote an AppleScript that manipulates the Photos export GUI dialog directly to get the full-quality edited files. [I posted the AppleScript file and a TypeScript wrapper to a gist](https://gist.github.com/kitschpatrol/d28fef0515fda20222dd7900ba76e365). Happy to share the whole comparison-generating script too, but it's in TypeScript, not Python, and kind of entangled in the rest of the stuff I'm working on.

Anyway, it's a quick hack, it's slow, and I'm finding it tricky to deduplicate the logic in AppleScript code. But it's nice to have all the other export options available — resizing, color profile handling, and so forth.

I also explored your (awesome) [PhotoScript](https://github.com/RhetTbull/photoscript) project, which definitely solves Photos-related AppleScript manipulation in a more robust way. I can contribute a PR integrating the export script if that's of interest, but honestly this approach is probably too brittle to be worth maintaining since it's just poking at the GUI instead of calling proper commands from the Photos automation dictionary. (Especially if the private APIs end up working instead!) 

Also FWIW, this round of testing confirmed that the "original" exports of unedited files are perfectly consistent across the range of export strategy. Not super surprising since it should just be a file copy, but at least there's nothing weird happening there.



And finally after futzing with the [Photo editing APIs](https://developer.apple.com/documentation/photokit/phasset/editing_asset_content), it seems like the promise of "lossless" editing across apps is elusive in practice, and the pipelines between Photos and other apps resemble a game of "bitmap potato" instead of thin stacks of constantly-recalculated adjustment metadata on top of the original image data. So far the only place I've ever seen an actual re-render of the edit stack kick in is through the Photos Export GUI...

The main issue seems to be that the "edited" JPEGs Photos.app saves to the Library are pretty heavily compressed, but Photos treats that as good enough to be the "highest quality" version of the edited file available via the public PhotoKit APIs...

The moral for me here was to avoid editing library photos across multiple apps, and definitely don't _start_ a stack of edits in Photos. Photomator, for example, saves its edited bitmaps at a higher quality. So, in theory working from the original Photos.app file in a third-party app first, and then making some final tweaks in Photos.app second is less destructive than going the other direction.

**Reply by kitschpatrol** [reply]
I just noticed a not-so-fun caveat: The export menu item in Photos does not preserve the alpha channel.

**Reply by RhetTbull** [reply]
@kitschpatrol 

> Digging into Photos.app a bit with Instruments, it seems to be calling a few methods of interest during the export process, things like:

I'm trying to use Instruments to debug another issue in Photos but I'm not very familiar with the tool. Any tips for how you configured it to see the calls that Photos was making?

**Reply by kitschpatrol** [reply]
Hi Rhet, I'm far from an Instruments expert, but can share my approach. I made a quick screen recording to show the steps. Basically cue up the action you want to analyze in Photos, then start the recording in Instruments, take the action, then stop the recording and dig into the various datapoints captured. There's not a ton of symbolication, so lots of calls are just memory addresses.

https://github.com/RhetTbull/osxphotos/assets/194164/3792f300-488f-4ae1-a0f1-2be2c325298e

**Reply by RhetTbull** [reply]
Thanks!

**Comment by kitschpatrol** [comment]
Hi Rhet, dusting off the project that had led to this thread earlier this year — everything's still working in Sequoia which is great.

Just wanted to offer one more data point... I was curious if Jpeg XL support might open up an improved export pipeline in Photos.app on Sequoia, but FWIW I'm still seeing alpha channel deletion when exporting Jpeg XL images with transparency to PNG/ TIFF from the export menu.

I'm not seeing any promising changes in the PhotoKit API diffs... any new insights from your end?
